fuel-compression/src/compaction.rs

```diff
@@ -0,0 +1,382 @@
+            .take_next();
+        <ChangesPerTable as access::AccessMut<T, WriteTo<T>>>::get_mut(&mut self.changes)
+            .values
+            .push(value);
+        Ok(key)
+    }
+}
+
+/// Convert data to reference-based format
+pub trait Compactable {
+    /// The compacted type with references
+    type Compact: Clone + Serialize + for<'a> Deserialize<'a>;
+
+    /// Count max number of each key type, for upper limit of overwritten keys
+    fn count(&self) -> CountPerTable;
+
+    /// Convert to compacted format
+    fn compact<R: RegistryDb>(
+        &self,
+        ctx: &mut CompactionContext<R>,
+    ) -> anyhow::Result<Self::Compact>;
+
+    /// Convert from compacted format
+    fn decompact<R: RegistryDb>(compact: Self::Compact, reg: &R) -> anyhow::Result<Self>
+    where
+        Self: Sized;
+}
+
+macro_rules! identity_compaction {
+    ($t:ty) => {
+        impl Compactable for $t {
+            type Compact = Self;
+
+            fn count(&self) -> CountPerTable {
+                CountPerTable::default()
+            }
+
+            fn compact<R: RegistryDb>(
+                &self,
+                _ctx: &mut CompactionContext<R>,
+            ) -> anyhow::Result<Self::Compact> {
+                Ok(*self)
+            }
+
+            fn decompact<R: RegistryDb>(
+                compact: Self::Compact,
+                _reg: &R,
+            ) -> anyhow::Result<Self> {
+                Ok(compact)
+            }
+        }
+    };
+}
+
+identity_compaction!(u8);
+identity_compaction!(u16);
+identity_compaction!(u32);
+identity_compaction!(u64);
+identity_compaction!(u128);
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct ArrayWrapper<const S: usize, T: Serialize + for<'a> Deserialize<'a>>(
+    #[serde(with = "serde_big_array::BigArray")] pub [T; S],
+);
+
+// TODO: use try_map when stabilized: https://github.com/rust-lang/rust/issues/79711
+#[allow(unsafe_code)]
+fn try_map_array<const S: usize, T, R, E, F: FnMut(T) -> Result<R, E>>(
+    arr: [T; S],
+    mut f: F,
+) -> Result<[R; S], E> {
+    // SAFETY: we are claiming to have initialized an array of `MaybeUninit`s,
+    // which do not require initialization.
+    let mut tmp: [MaybeUninit<R>; S] = unsafe { MaybeUninit::uninit().assume_init() };
+
+    // Dropping a `MaybeUninit` does nothing, so we can just overwrite the array.
+    for (i, v) in arr.into_iter().enumerate() {
+        tmp[i] = MaybeUninit::new(f(v)?);
+    }
+
+    // SAFETY: Everything element is initialized.
+    Ok(tmp.map(|v| unsafe { v.assume_init() }))
+}
+
+impl<const S: usize, T> Compactable for [T; S]
+where
+    T: Compactable + Clone + Serialize + for<'a> Deserialize<'a>,
+{
+    type Compact = ArrayWrapper<S, T::Compact>;
+
+    fn count(&self) -> CountPerTable {
+        let mut count = CountPerTable::default();
+        for item in self.iter() {
+            count += item.count();
+        }
```

Comment:
```suggestion
        self.iter().fold(count, |acc, item| acc + item.count());
```

Reply:
I find the plain loop way easier to read.